packages/coil-extension/src/background/services/Stream.ts

```diff
@@ -314,212 +319,138 @@ export class Stream extends EventEmitter {
-  private _active = true
-  private _lastDelivered = 0
-
-  constructor(opts: StreamAttemptOptions) {
-    this._onMoney = opts.onMoney
-    this._bandwidth = opts.bandwidth
-    this._plugin = opts.plugin
-    this._spspDetails = opts.spspDetails
-    this._debug = opts.debug
+class StreamLoop extends EventEmitter {
+  private state: StreamLoopState = StreamLoopState.Done
+  private readonly schedule: PaymentScheduler
+  private readonly debug: Logger
+  private running = false // whether there is an active loop
+  private connection?: Connection
+  constructor(opts: { schedule: PaymentScheduler; logger: Logger }) {
+    super()
+    this.schedule = opts.schedule
+    this.debug = opts.logger
   }
 
-  async start(): Promise<void> {
-    if (!this._active) return
-    const plugin = this._plugin
-
-    this._debug('creating ilp/stream connection.')
-    this._connection = await createConnection({
-      ...this._spspDetails,
-      plugin,
-      slippage: 1.0,
-      exchangeRate: 1.0,
-      maximumPacketAmount: '10000000',
-      getExpiry: getFarFutureExpiry
-    })
-
-    if (!this._active) return
-
-    // send practically forever at allowed bandwidth
-    this._debug('attempting to send on connection.')
-    this._ilpStream = this._connection.createStream()
-
-    // TODO: if we save the tier from earlier we don't need to do this async
-    // TODO: does doing this async allow a race condition if we stop right away
-    const initialSendAmount = await this._bandwidth.getStreamSendMax()
-    this._ilpStream.setSendMax(initialSendAmount)
-
-    return new Promise((resolve, reject) => {
-      const onMoney = (sentAmount: string) => {
-        // Wait until `setImmediate` so that `connection.totalDelivered` has been updated.
-        const receipt = this._ilpStream.receipt
-          ? this._ilpStream.receipt.toString('base64')
-          : undefined
-        setImmediate(this.onMoney.bind(this), sentAmount, receipt)
+  async run(
+    attempt: (tokenFraction: number) => Promise<Connection>
+  ): Promise<void> {
+    this.state = StreamLoopState.Loop
+    if (this.running) return // another 'run' loop is already active
+    this.running = true
+    this.schedule.start()
+    let attempts = 0 // count of retry attempts when state=Ending
+    while (this.state !== StreamLoopState.Done) {
+      let tokenPiece = 1.0
+      switch (+this.state) {
+        case StreamLoopState.Done:
+          continue // stop looping (unreachable)
+        case StreamLoopState.Loop: // keep going; only pay full tokens
+          // After the very first token payment, switch to post-payment.
+          if (0 < this.schedule.totalSent()) {
+            const hasToken = await this.schedule.awaitFullToken()
+            if (!hasToken) continue // the wait was interrupted, likely because the stream was paused/stopped
+          }
+          break
+        case StreamLoopState.Ending:
+          tokenPiece = Math.min(this.schedule.unpaidTokens(), 1.0)
```

Comment:
> after having prepaid a minute

Do you mean having paid/postpaid a minute? The extension only **pre**pays `INITIAL_SEND_SECONDS` (5 seconds), everything after that is post-payment.

When a stream is closed at 1.5 minutes, and 1 minute has already been paid, there is 30 seconds left to pay. So `this.schedule.unpaidTokens()` is 0.5, and it will pay half a token before ending.

Reply:
Right, yah. So this will only ever pay at most one token' worth when winding down? Would there ever be situations where the site is "owed" more than that?

Reply:
> So this will only ever pay at most one token' worth when winding down?

No. The `StreamLoop` will keep looping and paying until there are no remaining unpaid tokens.

Reply:
Ah, I see, yeah, it's just doing 1 at a time, but the the top while loop condition will keep it continuing ?

Reply:
Yes.